### ISSUE TITLE
Feat: Solidity Repl

### ISSUE THREAD

**Issue by brockelmore** [foundry-rs/foundry#230]
Have seen a few requests for a solidity repl. This can be a tracking issue for the feature

Helpful things:
 - https://github.com/tintinweb/solidity-shell
 - https://github.com/raineorshine/solidity-repl

**Comment by transmissions11** [comment]
Omg would be so dope if it had all your contract abis loaded to go so you could do something like `ERC20 token new ERC20()` in the shell right away and just play around with it

**Comment by mattsse** [comment]
some thoughts after skimming solidity-shell's codebase

it appears an appropriate approach to this could be:

1. run a repl like https://github.com/kkawakam/rustyline, see also https://github.com/kpcyrd/sn0int for an example implementation
2. listen for user input
   a. if it's a solidity statement, inject that into a template contract, compile and deploy that
   b. if it's some math expr, inject that into a function, deploy and return the response after calling that template function 
   c. else try to find a matching function that was previously deployed and call that

There's also [`TempProject`](https://github.com/gakonst/ethers-rs/blob/master/ethers-solc/src/project_util.rs#L13) now in ethers-solc which is probably helpful


If anyone would like to take this on, I'm happy to help

**Comment by transmissions11** [comment]
Can we persist variables? That’d be a dream

**Comment by mattsse** [comment]
totally, what I'm thinking is something like this:

```rust
pub struct SolRepl {
    /// this gives us the user input
    rl: Readline,
    /// this stores all previously evaluated user inputs, like vars, etc
    session: Session,
    /// the type that executes all the things
    sol: Sol,
}

struct Sol {
  /// the virtual project where we store all the things (perhaps not even necessary if we only use solc standard json input/output)
  project: TempProject,
  // stores all the deployed contracts/abis etc...
  state: SolState
}
```

then the actual repl would probably look something like:

```rust
loop {
 
 let input = self.rl.readline();
 let res = self.sol.execute(&line);
 println!("{}", res);
  }
```

**Comment by wilsoncusack** [comment]
Would love to see gas cost printed after evaluation :)

**Comment by mattsse** [comment]
closed via #3465
